src/MoBi.Presentation/Presenter/EditModuleConfigurationsPresenter.cs

```diff
@@ -0,0 +1,203 @@
+      }
+
+      private bool moduleInUse(Module module)
+      {
+         return _moduleConfigurations.Any(x => x.Uses(module));
+      }
+
+      private ModuleConfigurationDTO dtoFor(ModuleConfiguration moduleConfiguration)
+      {
+         return _moduleConfigurations.FirstOrDefault(x => x.Uses(moduleConfiguration));
+      }
+
+
+      private void addModuleToSelectionView(Module module)
+      {
+         _view.AddModuleNode(_treeNodeFactory.CreateFor(module));
+      }
+
+      public void AddModuleConfiguration(ITreeNode selectedTreeNode)
+      {
+         var selectedModule = projectModuleFor(selectedTreeNode);
+         if (selectedModule == null)
+            return;
+
+         var moduleConfiguration = new ModuleConfiguration(selectedModule);
+         var moduleConfigurationDTO = _moduleConfigurationDTOMapper.MapFrom(moduleConfiguration);
+         _moduleConfigurations.Add(moduleConfigurationDTO);
+         addModuleConfigurationToSelectedView(moduleConfiguration);
+         _view.RemoveNodeFromSelectionView(selectedTreeNode);
+      }
+
+      private void addModuleConfigurationToSelectedView(ModuleConfiguration moduleConfiguration)
+      {
+         _view.AddModuleConfigurationNode(_treeNodeFactory.CreateFor(moduleConfiguration));
+      }
+
+      public void RemoveModuleConfiguration(ITreeNode selectedTreeNode)
+      {
+         if (selectedTreeNode == null)
+            return;
+
+         var moduleConfigurationToRemove = moduleConfigurationDTOFor(selectedTreeNode);
+         _moduleConfigurations.Remove(moduleConfigurationToRemove);
+         _view.RemoveNodeFromSelectedView(selectedTreeNode);
+         addModuleToSelectionView(moduleConfigurationFor(selectedTreeNode).Module);
+      }
+
+      public void SelectedModuleConfigurationNodeChanged(ITreeNode selectedTreeNode)
+      {
+         var dto = moduleConfigurationDTOFor(selectedTreeNode);
+         _view.UnbindModuleConfiguration();
+
+         // if the user selects a non-module node we want disable the binding
+         if (dto == null)
+         {
+            _view.EnableRemove = false;
+            return;
+         }
+
+         _view.EnableRemove = true;
+         _view.BindModuleConfiguration(dto);
+      }
+
+      public void SelectedModuleNodeChanged(ITreeNode selectedNode)
+      {
+         var module = projectModuleFor(selectedNode);
+
+         _view.EnableAdd = module != null;
+      }
+
+      private Module projectModuleFor(ITreeNode selectedNode)
+      {
+         return selectedNode?.TagAsObject as Module;
+      }
+
+      private ModuleConfiguration moduleConfigurationFor(ITreeNode treeNode)
+      {
+         return treeNode?.TagAsObject as ModuleConfiguration;
+      }
+      
+      private ModuleConfigurationDTO moduleConfigurationDTOFor(ITreeNode treeNode)
+      {
+         var moduleConfiguration = moduleConfigurationFor(treeNode);
```

Comment:
It feels like we have 3 layers of indirection
The object, the node and the DTO. Why? (typically, a node is somekind of DTO)
it's hard to follow why goes on. What surprises me is that you are trying to retrieve. a DTO from an object. we use it sometimes but not so often

Reply:
The only thing that prevents me from getting rid of this DTO is that during configuration, I don't want to track changes using the domain object. The user could cancel, and I want to just get rid of everything. With the DTO I can keep track of selected start values easily. I think that's the only part where I need a temporary object besides the node. Although, the node also has child nodes with these objects as tags. I can try that.

Reply:
I can't tell. This is just hard to follow because of this level of indirection. In PKSim, we clone the whole config before editing. If the user cancels, we just discard. Otherwise we use as is. No need to synchronize anything

Reply:
A clone here will make everything easier.